classes/stream_wrapper.php

```diff
@@ -0,0 +1,542 @@
+ * https://github.com/aws/aws-sdk-php/blob/master/src/S3/StreamWrapper.php
+ * https://phpazure.codeplex.com/SourceControl/latest#trunk/library/Microsoft/WindowsAzure/Storage/Blob/Stream.php
+ *
+ * @package    local_azureblobstorage
+ * @author     Matthew Hilton <matthewhilton@catalyst-au.net>
+ * @copyright  Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class stream_wrapper {
+
+    /** @var resource|null Stream context (this is set by PHP) */
+    public $context;
+
+    /** @var StreamInterface Underlying stream resource */
+    private $body;
+
+    /** @var int Size of the body that is opened */
+    private $size;
+
+    /** @var string Mode in which the stream was opened */
+    private $mode;
+
+    /** @var string The opened protocol (e.g. "blob") */
+    private $protocol = 'blob';
+
+    /** @var HashContext Hash resource that is sent when flushing the file to Azure. */
+    private $hash;
+
+    /** @var bool records whether the file was readable when validating the stream_handle */
+    private $readable = true;
+
+    /** @var string The key extracted from the path from when the stream was opened */
+    private $key = null;
+
+    /**
+     * Register the blob://' stream wrapper
+     *
+     * @param api $client Client to use with the stream wrapper
+     * @param string $protocol Protocol to register as.
+     */
+    public static function register(api $client, $protocol = 'blob') {
+        if (in_array($protocol, stream_get_wrappers())) {
+            stream_wrapper_unregister($protocol);
+        }
+
+        stream_wrapper_register($protocol, get_called_class(), STREAM_IS_URL);
+        $default = stream_context_get_options(stream_context_get_default());
+        $default[$protocol]['client'] = $client;
+        stream_context_set_default($default);
+    }
+
+    /**
+     * Does not support casting, always returns false.
+     * @param mixed $castas
+     * @return bool
+     */
+    public function stream_cast($castas): bool {
+        return false;
+    }
+
+    /**
+     * Closes the stream
+     */
+    public function stream_close() {
+        $this->body = null;
+        $this->hash = null;
+    }
+
+    /**
+     * Opens the stream, depending on the mode.
+     * @param mixed $path filepath
+     * @param mixed $mode file mode constants, see fopen
+     * @param mixed $options Additional flags
+     * @param mixed $openedpath
+     * @return bool True if successful, else false
+     * @see https://www.php.net/manual/en/function.fopen.php
+     */
+    public function stream_open($path, $mode, $options, &$openedpath): bool {
+        // Set the protocol.
+        $this->initProtocol($path);
+        $this->key = $this->get_key_from_path($path);
+
+        // Trim 'b' and 't' off end.
+        // these are line ending flags used to handle Unix/Windows line endings.
+        // We don't care about these, so we just remove them.
+        $this->mode = rtrim($mode, 'bt');
```

Comment:
Does this break compatibility with windows at all?

Reply:
Interesting thought - had a bit of a look into it and the answer is likely no - because object stores always use binary (the b mode) and don't do line ending conversion (the t mode), and also the AWS stream wrapper also trims these off https://github.com/catalyst/moodle-local_aws/blob/master/sdk/Aws/S3/StreamWrapper.php#L152